Add TLS support to exec authenticator plugin

awly opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/kubectl-exec-plugins.md#tls-client-certificate-support

Allows exec plugin to return raw TLS key/cert data. This data populates
transport.Config.TLS field.
This requires a change to AuthProvider interface to expose TLS configs,
not only RoundTripper.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61421

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Exec authenticator plugin supports TLS client certificates.
```

ericchiang commented:
/ok-to-test

awly commented:
/retest

make bazel-test passes locally
make bazel-test-integration fails at //test/integration/master:go_default_test on this PC and parent commit

ericchiang commented on staging/src/k8s.io/client-go/rest/plugin.go:
```diff
@@ -18,18 +18,23 @@ package rest
 type AuthProvider interface {
-	// WrapTransport allows the plugin to create a modified RoundTripper that
-	// attaches authorization headers (or other info) to requests.
-	WrapTransport(http.RoundTripper) http.RoundTripper
```
I think you can make this change without changing the interface since the exec provider doesn't use it. If you want to keep the diff smaller.

awly replied:
Good point, reverted interface changes

ericchiang commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -86,22 +86,22 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
 	}
 
-	c.TLS.CertData, err = dataFromSliceOrFile(c.TLS.CertData, c.TLS.CertFile)
+	c.TLS.CertData, err = dataFromSliceFuncOrFile(c.TLS.GetCert, c.TLS.CertData, c.TLS.CertFile)
```
I expected using [`crypto/tls.Config.GetClientCertificate`](https://golang.org/pkg/crypto/tls/#Config) instead. That way you could actually rotate these on a live transport.

mikedanese replied:
I think we also need to close old connections when a rotation occurs. See:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/certificate/transport.go#L149

awly replied:
Implemented GetClientCertificate.
Will work on getting old connections closed on cert rotation next.

ericchiang commented:
This looks like a real failure

```
--- FAIL: TestTLSCredentials (0.05s)
    --- FAIL: TestTLSCredentials/valid_TLS_cert (0.00s)
    	exec_test.go:508: TLSConfigFor: decode stdout: couldn't get version/kind; json parse error: invalid character '\n' in string literal
    --- FAIL: TestTLSCredentials/untrusted_TLS_cert (0.00s)
    	exec_test.go:508: TLSConfigFor: decode stdout: couldn't get version/kind; json parse error: invalid character '\n' in string literal
    --- FAIL: TestTLSCredentials/valid_TLS_cert_again (0.00s)
    	exec_test.go:508: TLSConfigFor: decode stdout: couldn't get version/kind; json parse error: invalid character '\n' in string literal
FAIL
```

awly commented:
Added connection rotation on cert change, PTAL

awly commented:
Tests fixed (thanks @mikedanese for help).

mikedanese commented on staging/src/k8s.io/client-go/util/conntrack/conntrack.go:
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

awly replied:
Done

awly commented:
/retest

mikedanese commented on staging/src/k8s.io/client-go/transport/transport_test.go:
```diff
@@ -168,37 +173,114 @@ func TestNew(t *testing.T) {
-				continue
+				t.Fatalf("got %#v, expected no TLSClientConfig", transport)
+			}
+			if testCase.TLS {
```
Here and for TLSCert, do like you do above:
```golang
if !testCase.TLS{
  return
}
```

awly replied:
Done

mikedanese reviewed (commented):
looks pretty good, can you squash fixup commits?

awly commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -250,31 +314,49 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
+		token: cred.Status.Token,
+	}
+
+	// Avoid setting non-nil key/cert when plugin didn't return them.
```
This would happen if exec plugin provides tokens.
Since a.cachedCreds.key/cert are []byte fields, i want to keep them nil instead of []byte{} when plugin didn't return key/cert. This just makes checks cleaner everywhere (foo == nil vs len(foo) == 0)

Clarified that it's about []byte{} in the comment

awly commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -191,40 +229,66 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	return a.now().After(a.exp)
 }
 
 func (a *Authenticator) token() (string, error) {
```
Done

ericchiang commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -191,40 +229,66 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return creds.token, nil
+}
+
+func (a *Authenticator) cert() ([]byte, error) {
```
This has the potential to race with key()
```
cert := a.cert()
// key could rotate here
key := a.key()
```

awly replied:
Nice catch. Merged GetCert/GetKey and cert/key here.

ericchiang commented on staging/src/k8s.io/client-go/transport/config.go:
```diff
@@ -98,4 +103,7 @@ type TLSConfig struct {
 	CertData []byte // Bytes of the PEM-encoded client certificate. Supercedes CertFile.
 	KeyData  []byte // Bytes of the PEM-encoded client key. Supercedes KeyFile.
+
+	GetCert func() ([]byte, error) // Callback that returns PEM-encoded client certificate. Supercedes CertData and CertFile.
```
Maybe just:

```
GetCert func() (*tls.Certificate, error)
```

awly replied:
Done

ericchiang commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -250,31 +314,49 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
+	}
+
+	if a.onRotate != nil && !reflect.DeepEqual(oldCreds, a.cachedCreds) {
+		a.onRotate()
```
We don't need to rotate if the Token changes but the cert doesn't. Maybe a TODO if it's hard to change right now.

awly replied:
Done

awly commented:
PTAL

ericchiang commented on staging/src/k8s.io/client-go/util/conntrack/conntrack.go:
```diff
@@ -0,0 +1,102 @@
+
+// Package conntrack implements a connection dialer that tracks and can close
+// all created connections.
+package conntrack
```
It'd be great to make this internal, but that won't be easy between `k8s.io/kubernetes` and `k8s.io/client-go`.

lavalamp replied:
Please pick a different name.

ericchiang commented on staging/src/k8s.io/client-go/transport/cache.go:
```diff
@@ -51,7 +52,7 @@ func (t tlsCacheKey) String() string {
 		keyText = "<redacted>"
 	}
-	return fmt.Sprintf("insecure:%v, caData:%#v, certData:%#v, keyData:%s, serverName:%s", t.insecure, t.caData, t.certData, keyText, t.serverName)
+	return fmt.Sprintf("insecure:%v, caData:%#v, certData:%#v, keyData:%s, getCert: %s, serverName:%s", t.insecure, t.caData, t.certData, keyText, t.getCert, t.serverName)
```
Hmmm, how does this work with getCert?

awly replied:
It uses the function address to uniquely identify the cached http.Transport.
It's not really calling GetCert.

ericchiang commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -191,40 +228,58 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	defer a.mu.Unlock()
 
-	if token != a.cachedToken {
+	if token != a.cachedCreds.token {
```
We need to document that 401s wont cause client certs to rotate. I guess we don't have the ability to revoke certs right now so it's not that big of a deal?

awly replied:
Invalid client certs should fail at TLS handshake. There won't be an HTTP response.
And tls.Dial doesn't seem to provide a way to differentiate between "bad client cert" vs any regular network error?
Exec plugin could also do some cert caching, so re-execing it doesn't guarantee rotation.

I propose to leave this as is and maybe recommend short-lived certs?

ericchiang replied:
Ah of course. I don't think we should worry about this for now.

mikedanese replied:
i think it's possible to get a 401 if the connection was established before the client cert expired but the cert is now expired.

awly replied:
Does the server check client cert after handshake at all?
If yes and this edge case gets hit too often, exec plugin can simply return expirationTime = certExpirationTime - 1d. That should rotate all connections before cert expiry.

liggitt replied:
> Invalid client certs should fail at TLS handshake. There won't be an HTTP response.

that's not accurate... because the apiserver allows for both TLS and token auth, it does the client cert issuer/expiry/key-usage check post-handshake. only private key proof-of-possession checks are done during the handshake.

ericchiang reviewed (commented):
couple nits, other than that lgtm

awly commented:
/retest

mikedanese commented:
/lgtm

awly commented:
/assign @lavalamp

awly commented:
ping @lavalamp

awly commented:
/hold
Just tried it on a real cluster and getting a panic
runtime: goroutine stack exceeds 1000000000-byte limit

Probably something in conntrack lib. Debugging...

ericchiang commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -147,14 +153,41 @@ type Authenticator struct {
+
+	defaultDialer := &net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second}
+	ct := conntrack.NewTracker(func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if c.Dial != nil {
```
I think this is your bug. You don't save the dialer before setting it later like you do with the WrapTransport.
```
dial := c.Dial

ct := conntrack.NewTracker(.... {
    if dial != nil {
        return dial(network, addr)
    }
    return defaultDialer.DialContext(ctx, network, addr)
})
```

awly replied:
Yep, you're right. That was it.

ericchiang commented:
/retest
/lgtm

awly commented:
/unassign @lavalamp 
/assign @smarterclayton

lavalamp commented on staging/src/k8s.io/client-go/util/conntrack/conntrack.go:
```diff
@@ -0,0 +1,102 @@
+limitations under the License.
+*/
+
+// Package conntrack implements a connection dialer that tracks and can close
```
Please document: Why would I want to do this? What problem does this solve?

Consider putting the package comment in a doc.go file. (not mandatory since this seems to be a one file package)

awly replied:
Added a bit more on purpose of the package.

lavalamp commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/types.go:
```diff
@@ -58,6 +58,9 @@ type ExecCredentialStatus struct {
 	Token string `json:"token,omitempty"`
+	// PEM-encoded client TLS certificate and key.
+	ClientCertificateData string `json:"clientCertificateData,omitempty"`
+	ClientKeyData         string `json:"clientKeyData,omitempty"`
```
Is this a private key? Can you explain in the documentation why it is safe to transmit a private key?

awly replied:
Yes. Added a comment to wrapper struct.
This is not a message that will get transmitted over the network. It's for exec plugins (same host). The whole purpose of it is to generate key/cert and give it to client (kubelet).

lavalamp commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/types.go:
```diff
@@ -58,6 +58,9 @@ type ExecCredentialStatus struct {
 	ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
 	// Token is a bearer token used by the client for request authentication.
 	Token string `json:"token,omitempty"`
+	// PEM-encoded client TLS certificate and key.
```
Document both fields separately so the generated documentation makes sense. Document what this key is used for (who is the client and who is the server). Our certs, cas, and keys are VERY confusing already. What may be in this certificate data? Is it a ca bundle? A single cert?

awly replied:
Split comments and elaborated on cert contents

awly commented:
Thanks for review @lavalamp. PTAL.

ericchiang commented:
Still need to rename the `conntrack` package https://github.com/kubernetes/kubernetes/pull/61803#discussion_r181244219

Maybe we can move it's contents to client-go's transport package?

https://godoc.org/k8s.io/client-go/transport

awly commented:
Oh, not sure how I missed that comment. Thanks @ericchiang !

Renamed to `connreset` and squashed fixup commits. If you can think of a better name, let me know.

I think https://godoc.org/k8s.io/client-go/transport is not a good place for this.
That package looks entirely HTTP-centric (all about `http.RoundTrippers` and pretty sure it's named after `http.Transport`). `connreset` is just TCP.

mikedanese commented:
/retest

awly commented:
/retest

lavalamp commented on staging/src/k8s.io/client-go/util/connreset/connreset.go:
```diff
@@ -0,0 +1,105 @@
+//
+// This is used for credential rotation of long-lived connections, when there's
+// no way to re-authenticate on a live connection.
+package connreset
```
I suggest some combo of "connection" and "rotate", given the documentation. or something like socketrotation, dialerrotation, etc

lavalamp replied:
BTW I strongly recommend in the future if you do something like this that you do the refactoring and the new feature implementation in 100% separate PRs. Those PRs would be much easier to review.

lavalamp reviewed (commented):
I have one suggestion, I guess it is optional. Someone from sig auth should review this.

mikedanese commented:
@kubernetes/sig-auth-pr-reviews 

@liggitt might be a good approver if he has time since he approved the first part of this.

tallclair commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/types.go:
```diff
@@ -52,12 +52,20 @@ type ExecCredentialSpec struct {
 
 // ExecCredentialStatus holds credentials for the transport to use.
+//
+// Most of the fields below are sensitive. This data should only be transmitted
```
This comment only applies to the `Token` & `ClientKeyData` fields, right?

awly replied:
Yep, clarified

tallclair commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -147,14 +153,44 @@ type Authenticator struct {
+	// c.Dial->ct.Dial->c.Dial.
+	dial := c.Dial
+	d := connrotation.NewDialer(func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if dial != nil {
```
nit: this wouldn't change, right? would it make sense to move this out of this function? e.g.
```go
var dial func(ctx context.Context, network, addr string) (net.Conn, error)
if c.Dial != nil {
  dial = c.Dial
} else {
  dial = &net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second}
}
```

awly replied:
Good point. It's a bit uglier than your snipped, but avoids an extra branch after setup

tallclair commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -76,11 +76,37 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
+					return cert, nil
+				}
+			}
+			// Both c.TLS.CertData/KeyData were unset and GetCert didn't return
```
Comment doesn't match logic. Both ... were unset implies `&&`.

awly replied:
Fixed

awly commented:
ping @smarterclayton

liggitt commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1/types.go:
```diff
@@ -52,12 +52,20 @@ type ExecCredentialSpec struct {
 	ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
 	// Token is a bearer token used by the client for request authentication.
 	Token string `json:"token,omitempty"`
+	// PEM-encoded client TLS certificate (without CA data).
```
in some cases, intermediate CA data might be required. no reason to require this to omit that... we should handle a blob containing 1 or more certificate PEM blocks just like golang, which takes the first as the client cert and the rest as the intermediate pool to use

awly replied:
Done

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -250,31 +308,51 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
 	_, gvk, err := codecs.UniversalDecoder(a.group).Decode(stdout.Bytes(), nil, cred)
 	if err != nil {
-		return "", fmt.Errorf("decode stdout: %v", err)
+		return fmt.Errorf("decode stdout (%q): %v", stdout.String(), err)
```
not sure we want to print content likely to contain credentials

awly replied:
Removed output body

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -250,31 +308,51 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
 
-	return a.cachedToken, nil
+	oldCreds := a.cachedCreds
+	a.cachedCreds = &credentials{
```
probably don't want to stomp a.cachedCreds until after we've finished parsing the response from the exec plugin below (once we're sure we're not going to return an error from this method)

awly replied:
Done

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -191,40 +231,58 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 }
 
 func (a *Authenticator) token() (string, error) {
+	creds, err := a.freshCreds()
```
the name freshCreds() makes me think it is skipping the cache...

awly replied:
Renamed a few methods to make it clearer

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -173,6 +209,10 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	if err != nil {
```
this should switch to getting the whole creds set, and passing the whole creds set to `refresh()`

awly replied:
Done

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -173,6 +209,10 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	if err != nil {
 		return nil, fmt.Errorf("getting token: %v", err)
 	}
+	// The plugin only provided TLS credentials.
```
are we not checking for a refresh on 401 responses for client certs?

awly replied:
We are now, I didn't think TLS handshake would succeed on bad certs.

liggitt commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -76,11 +76,37 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
-		if err != nil {
+		// If key/cert were provided, verify them before setting up
+		// tlsConfig.GetClientCertificate.
+		if _, err := tls.X509KeyPair(c.TLS.CertData, c.TLS.KeyData); err != nil {
```
save the result of this, and return that from GetClientCertificate as the fallback case rather than reparsing the certificate on every call to GetClientCertificate

awly replied:
Done

awly commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -76,11 +76,37 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
+			// Both c.TLS.CertData/KeyData were unset and GetCert didn't return
+			// anything. Don't provide client cert to server.
+			if len(c.TLS.CertData) == 0 && len(c.TLS.KeyData) == 0 {
+				return nil, nil
```
Nice catch, done

awly commented:
PTAL @liggitt 
Made a separate comment so it's easier to see the diff. Will squash later.

awly commented:
Adding Jordan as approver to make sure the changes LGTY
/assign @liggitt

awly commented:
ping @smarterclayton

awly commented:
ping @liggitt @smarterclayton

awly commented:
/retest

awly commented:
ping @liggitt @smarterclayton

tallclair commented:
Can you split the refactor portion of this PR (moving connTracker) into a separate PR? I think that will make both easier to review, and should help accelerate the approval process. Thanks!

awly commented:
Split out connection rotation refactor into https://github.com/kubernetes/kubernetes/pull/63936

awly commented:
@mikedanese @ericchiang PTAL, rebased over https://github.com/kubernetes/kubernetes/pull/63936

awly commented:
/unassign @smarterclayton

mikedanese commented:
I'm still lgtm on this.
/lgtm

awly commented:
ping @liggitt

awly commented:
ping @liggitt

awly commented:
/retest

liggitt commented:
Apologies for the latency. This is on my list to review this week. One question I had was around override behavior when credentials were already present. Previously, when a token was passed, the auth provider would honor it / not stomp the token. Is there similar behavior when an existing client cert is present? Is the behavior to avoid overriding the token maintained with this PR?

awly commented:
Good point.
Just updated it to give precedence to static key/cert in `transport.Config` over `GetCert` callback and to existing callback over exec plugin.

Token behavior is unchanged.

awly commented:
ping @liggitt

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -184,47 +233,60 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-func (a *Authenticator) getToken(r *clientauthentication.Response) (string, error) {
+// refreshCreds executes the plugin and reads the credentials from stdout. It
+// must be called while holding the Authenticator's mutex.
+func (a *Authenticator) refreshCreds(r *clientauthentication.Response) error {
```
nit: rename to refreshCreds_locked to make sure callers know they must hold the lock

awly replied:
Done (except refreshCredsLocked)

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -250,31 +312,51 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
 	if cred.Status.ExpirationTimestamp != nil {
 		a.exp = cred.Status.ExpirationTimestamp.Time
 	} else {
 		a.exp = time.Time{}
```
if no expiration was returned, and we have a certificate, do we want to compute an expiration time from the certificate Not After timestamp?

awly replied:
I think that logic should live in the plugin itself.
This simplifies the contract between caller and plugin and makes it easier to reason about.

liggitt replied:
that's probably fine to start. if we find it is burdensome for plugins to pull this info out of the cert, we can revisit

liggitt commented on staging/src/k8s.io/client-go/rest/transport.go:
```diff
@@ -111,5 +82,29 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
+			return nil, err
+		}
+	}
+	if c.AuthProvider != nil {
```
@ericchiang remind me the use case for setting both AuthProvider and ExecProvider? was there a reason we didn't make those exclusive when adding ExecProvider in 1.10?

ericchiang replied:
If they're not mutually exclusive that wasn't intended. I wouldn't expect you to be able to provide AuthProvider and ExecProvider.

Maybe I should change that as part of the beta promotion?

awly replied:
I agree, this should be disallowed as part of beta promotion PR

liggitt commented on staging/src/k8s.io/client-go/transport/config.go:
```diff
@@ -99,4 +105,6 @@ type TLSConfig struct {
 	CertData []byte // Bytes of the PEM-encoded client certificate. Supercedes CertFile.
 	KeyData  []byte // Bytes of the PEM-encoded client key. Supercedes KeyFile.
+
+	GetCert func() (*tls.Certificate, error) // Callback that returns a TLS client certificate. Supercedes CertData, CertFile, KeyData and KeyFile.
```
@kubernetes/sig-api-machinery-pr-reviews this makes transport.TLSConfig non-serializable. From what I can see, it is only used in transport.Config, which is also non-serializable (has WrapTransport, Dial, etc), so I don't think that's a problem, but wanted to call it out

liggitt replied:
the comment about "supercedes certdata, ..." isn't accurate any more. see `Note: static key/cert data always take precedence over cert callback.` comment below

awly replied:
Updated comment

liggitt commented on staging/src/k8s.io/client-go/transport/cache.go:
```diff
@@ -109,6 +110,7 @@ func tlsConfigKey(c *Config) (tlsCacheKey, error) {
 		caData:     string(c.TLS.CAData),
 		certData:   string(c.TLS.CertData),
 		keyData:    string(c.TLS.KeyData),
+		getCert:    fmt.Sprint(c.TLS.GetCert),
```
what does this print as? I expected %p like we use for c.Dial

liggitt replied:
also, if you're changing this, update TestTLSConfigKey to test for uniqueness

awly replied:
`fmt.Sprint` on functions is same as `%p`. Changed for consistency.
Updates the test too.

awly commented on staging/src/k8s.io/client-go/pkg/apis/clientauthentication/types.go:
```diff
@@ -58,6 +58,9 @@ type ExecCredentialStatus struct {
 	Token string
+	// PEM-encoded client TLS certificate and key.
+	ClientCertificateData string
+	ClientKeyData         string
```
Done

ericchiang commented:
/test pull-kubernetes-e2e-kops-aws

Commit:
Add TLS support to exec authenticator plugin

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/kubectl-exec-plugins.md#tls-client-certificate-support

Allows exec plugin to return raw TLS key/cert data. This data populates
transport.Config.TLS fields.
transport.Config.TLS propagates custom credentials using
tls.Config.GetClientCertificate callback.
On key/cert rotation, all connections using old credentials are
closed

liggitt commented:
/lgtm
/status approved-for-milestone

liggitt commented:
@awly can you open a doc PR against the release-1.11 branch of https://github.com/kubernetes/website updating the doc for https://github.com/kubernetes/website/blob/release-1.11/content/en/docs/reference/access-authn-authz/authentication.md#input-and-output-formats with info about the TLS response

liggitt commented:
/assign @cblecker @sttts 
for approval of added godep package dependency

cblecker commented:
All internal packages
/approve

liggitt commented:
/retest

awly commented:
@liggitt sent docs PR https://github.com/kubernetes/website/pull/8826

awly commented:
/retest
